cmd/getStorageValue.go

```diff
@@ -0,0 +1,151 @@
+// limitations under the License.
+
+package cmd
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/makerdao/vulcanizedb/libraries/shared/factories/storage"
+	storage2 "github.com/makerdao/vulcanizedb/libraries/shared/storage"
+	"github.com/makerdao/vulcanizedb/libraries/shared/storage/types"
+	"github.com/makerdao/vulcanizedb/libraries/shared/transformer"
+	"github.com/makerdao/vulcanizedb/pkg/core"
+	"github.com/makerdao/vulcanizedb/pkg/datastore/postgres"
+	"github.com/makerdao/vulcanizedb/pkg/datastore/postgres/repositories"
+	"github.com/makerdao/vulcanizedb/utils"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var getStorageValueBlockNumber int64
+
+// getStorageValueCmd represents the getStorageValue command
+var getStorageValueCmd = &cobra.Command{
+	Use:   "getStorageValue",
+	Short: "Gets all storage values for configured contracts at the given block.",
+	Long: `Fetches and persists storage values of the configured contracts at a given block. It is important to note that the storage value gotten with this
+	command may not be different from the previous block in the database.`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		SubCommand = cmd.CalledAs()
+		LogWithCommand = *logrus.WithField("SubCommand", SubCommand)
+		fmt.Println("getStorageValue called")
+		return getStorageAt(getStorageValueBlockNumber)
+	},
+}
+
+func init() {
+	getStorageValueCmd.Flags().Int64VarP(&getStorageValueBlockNumber, "get-storage-value-block-number", "b", -1, "block number to fetch storage at for all configured transformers")
+	rootCmd.AddCommand(getStorageValueCmd)
+}
+
+func getStorageAt(blockNumber int64) error {
+	blockChain := getBlockChain()
+	db := utils.LoadPostgres(databaseConfig, blockChain.Node())
+	storageInitializers := exportTransformers()
+	commandRunner := NewStorageValueCommandRunner(blockChain, &db, storageInitializers, blockNumber)
+	return commandRunner.Run()
+}
+
+func NewStorageValueCommandRunner(bc core.BlockChain, db *postgres.DB, initializers []transformer.StorageTransformerInitializer, blockNumber int64) StorageValueCommandRunner {
+	return StorageValueCommandRunner{
+		bc:              bc,
+		db:              db,
+		headerRepo:      repositories.NewHeaderRepository(db),
+		storageDiffRepo: storage2.NewDiffRepository(db),
+		initializers:    initializers,
+		blockNumber:     blockNumber,
+	}
+}
+
+type StorageValueCommandRunner struct {
+	bc              core.BlockChain
+	db              *postgres.DB
+	headerRepo      repositories.HeaderRepository
+	storageDiffRepo storage2.DiffRepository
+	initializers    []transformer.StorageTransformerInitializer
+	blockNumber     int64
+}
+
+func (r *StorageValueCommandRunner) Run() error {
+	addressToKeys, getKeysErr := r.getStorageKeys()
+	if getKeysErr != nil {
+		return getKeysErr
+	}
+
+	header, getHeaderErr := r.headerRepo.GetHeader(r.blockNumber)
```

Comment:
Currently this returns an error if we don't have a header for the given block in the db yet. I _think_ that should be okay, because we'll likely not run this command super close to the head of the chain, but I'd be curious to hear what folks think about that. The other idea I had was to get the header (which we need to get the hash for the storage_diff table) directly from the blockchain, but thought that saving ourselves another rpc call may be worth it, especially since we already should have the header in the db.